### PR TITLE
Keycloak Authentication

### PR DESCRIPTION
**Pull request by HLWeil**
Added and modified several commands related to authentication to git services:

- `arc auth` will open a browser window of a token specified in the arc config. The user can login there and a the credentials to a specific git service (like a gitlab instance) will be automatically stored using the git credential manager interface. The response from the token service additionally contains user metadata (`name` and `email`) used in signing commits. These will be stored in the arc config. *(Command name deprecated, see [this comment](https://github.com/nfdi4plants/arcCommander/pull/103#issuecomment-1034787000))*
- `arc sync` now checks whether this user metadata in the arc config matches the metadata in the git config. If not, it will abort and suggest to the user to run the following command:
- `arc config setgituser` will take the git user metadata stored in the arc config and transfer it to the git config. Per default only the local git config of the repo will be overwritten. But this can be changed via `--local` and `--global` flags. Additionally, alternative `name` and `email` can also be provided

So a common workflow will look like this: 

A user must use `arc auth` once for each distinct gitlab instance he wants access to. Then they can download arcs from that gitlab using `arc get`. If they want to store their git user metadata for each arc locally they then have to run `arc config setgituser`. Alternatively they can run it once globally after authentication. Afterwards they can use `arc sync` everytime they want to pull or push changes for this arc.

Edit: Pronouns
Edit2: Added reference to comment showing new command syntax

### CODE REVIEW

**Comment by chgarth** [comment]
Interesting work! Could you elaborate briefly what the intended use cases are? It almost reads as if direct auth with some GitLab instance is now the preferred way of accessing ARCs, which I think should not be the goal, due to breaking the prior constraint of not taking advantage of any specific implementation (could have a GitHub repo or some other Git host).

I do think it's ok to have functions in the ARC commander to help users set up SSH keys / access tokens in specific circumstances, e.g. for DataPLANT HUB. From a security POV, I don't think it's wise to store user credentials inside an ARC?

---

Mini-nitpick: not all users are male, suggest using "they/them" as pronouns.

File: src/ArcCommander/APIs/ConfigurationAPI.fs [hunk 1]
```diff
@@ -173,4 +173,47 @@ module ConfigurationAPI =
...
+    let setGitUser (arcConfiguration : ArcConfiguration) (configurationArgs : Map<string,Argument>) =
+
+        let log = Logging.createLogger "ConfigurationSetGitUserLog"
+
+        log.Info("Start Config SetGitUser")
+
+        let nameOption = 
+            match tryGetFieldValueByName "Name" configurationArgs with
+            | Some name -> 
+                log.Trace("TRACE: Retrieved user name from given argument.")
```
Comment (omaus) [inline src/ArcCommander/APIs/ConfigurationAPI.fs:189]: "TRACE: " prefix is not used. The only prefixes used are in `log.Warn` ("WARNING: "), `log.Error`, and `log.Fatal` ("ERROR: ")

File: src/ArcCommander/APIs/ConfigurationAPI.fs [hunk 2]
```diff
@@ -173,4 +173,47 @@ module ConfigurationAPI =
...
+                
+            if (containsFlag "Global" configurationArgs |> not) || containsFlag "Local" configurationArgs then
+                let workDir = GeneralConfiguration.getWorkDirectory arcConfiguration
+                GitHelper.setLocalName workDir name
+                GitHelper.setLocalEmail workDir email
+
+            log.Info("Finished setting git user information.")
+
+        | _, _ -> 
+            log.Error("ERROR: Git user metadata neither present in the arc config nor given by argument. Consider first running \"arc auth\", setting the user name and email by running \"arc config set\" or rerunning this command and specifying both name and email.")
```
Comment (omaus) [inline src/ArcCommander/APIs/ConfigurationAPI.fs:219]: "[...] present in the ***ARC*** config [...]". "arc" only when referring to commands: "`arc auth`" etc.

File: src/ArcCommander/APIs/GitAPI.fs [hunk 3]
```diff
@@ -25,6 +17,40 @@ module GitAPI =
...
+    /// Authenticates to a token service and stores the token using git credential manager.
+    let authenticate (arcConfiguration : ArcConfiguration) =
+
+        let log = Logging.createLogger "GitAuthenticateLog"
+
+        log.Info("Start Arc Authenticate")
+        
+        match Authentication.tryLogin log arcConfiguration with 
+        | Ok token -> 
+            log.Info($"INFO: Successfully retrieved access token from token service")
```
Comment (omaus) [inline src/ArcCommander/APIs/GitAPI.fs:29]: You don't need interpolated strings when nothing is interpolated. 😄 
Also, no "INFO: " prefix.
-> 
```f#
log.Info("Successfully retrieved access token from token service")
```

File: src/ArcCommander/APIs/GitAPI.fs [hunk 4]
```diff
@@ -25,6 +17,40 @@ module GitAPI =
...
+                IniData.setValueInIniPath globalConfigPath "general.gitname"    (token.FirstName + " " + token.LastName)
+                IniData.setValueInIniPath globalConfigPath "general.gitemail"   token.Email
+                log.Trace($"TRACE: Successfully transferred git user metadata to global arcCommander config")
+            | None ->
+                log.Error($"ERROR: Could not transfer git user metadata to global arcCommander config")
+
+            if storeCredentialsToken log token then
+                log.Info($"Finished Authentication")
+
+            else
```
Comment (omaus) [inline src/ArcCommander/APIs/GitAPI.fs:43]: What is the reason for using several log messages? For normal, we use one log message for one event:
```f#
log.Error("ERROR: Authentication worked, but credentials could not be stored successfully.\nCheck if git is installed and if a credential helper is setup:\nRun \"git config --global credential.helper cache\" to cache credentials in memory\nor Run \"git config --global credential.helper store\" to save credentials to disk\nFor more info go to: https://git-scm.com/book/en/v2/Git-Tools-Credential-Storage")
```

File: src/ArcCommander/ArcCommander.fsproj [hunk 5]
```diff
@@ -4,7 +4,7 @@
     <OutputType>Exe</OutputType>
     <IncludeAllContentForSelfExtract>true</IncludeAllContentForSelfExtract>
     <TargetFramework>net6.0</TargetFramework>
-    <CopyLocalLockFileAssemblies>false</CopyLocalLockFileAssemblies>
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
```
Comment (omaus) [inline src/ArcCommander/ArcCommander.fsproj:7]: Why is this?

File: src/ArcCommander/Authentication.fs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+            //        .CreateLogger()
+    
+            //options.LoggerFactory.AddSerilog(serilog) |> ignore
+
+            log.Trace($"TRACE: Prepare client for login procedure")
+
+            let client = new OidcClient(options)
+            let! state = client.PrepareLoginAsync()
+    
+            //log.Trace($"TRACE: Client setup")
```
Comment (omaus) [inline src/ArcCommander/Authentication.fs:140]: Delete?

**Comment by HLWeil** [comment]
Thanks for your input!

> Interesting work! Could you elaborate briefly what the intended use cases are? It almost reads as if direct auth with some GitLab instance is now the preferred way of accessing ARCs, which I think should not be the goal, due to breaking the prior constraint of not taking advantage of any specific implementation (could have a GitHub repo or some other Git host).

The `arc auth` functionality is not meant to directly connect to a `GitLab` instance. Sry that was a bit missworded. The goal here is to receive an access token from an `authorization service`. This will enable _hopefully_ seemless `authorization` for the specific `git service` this token is meant to be used for. 
The `arc sync` and `arc get` are not necessarily dependent on this. One could directly use these commands on e.g. an `arc` stored in `Github` or an `GitLab` instance.

> I do think it's ok to have functions in the ARC commander to help users set up SSH keys / access tokens in specific circumstances, e.g. for DataPLANT HUB. From a security POV, I don't think it's wise to store user credentials inside an ARC?

The user `name` and `password` are only used in the browser window of the `token service`. The received `username` and `access token`  for the git service then get handed `directly to git` and not stored in the arc config. Stored in the arc config is only the the metainformation used to sign commits: `Full name` and `e-mail` of the user. But that is something anyone can see if he checks git locally or visits the git service online.

> Mini-nitpick: not all users are male, suggest using "they/them" as pronouns.

Done ;)

**Comment by HLWeil** [comment]
New commit: Moved `arc auth` into a command tree for handling remote access to make its name more descriptive of what it actually does. The command now also takes the `authentication protocol` and the `host address` as arguments. E.g.:

`arc remoteaccess accesstoken get --openid -s "https://keycloak.dev.bwsfs.uni-freiburg.de/auth/realms/sandbox"`

I'm currently working on functionality to receive tokens via the `oauth2` authentication protocol, which `Github` uses. This will then hopefully look like this: 

`arc remoteaccess accesstoken get --oauth2 -s "https://GitHub.com"`

**Comment by HLWeil** [comment]
Will merge this for now, receiving access token trough `oauth2` will be added in a later PR

**Pull request merged**